New: Add maximum single episode age option (per indexer)

cjmanca opened the pull request:
#### Description
Option to avoid grabbing single episode releases for older episodes. This still allows it to get new episodes as they come out, but only allows season packs for older seasons.

Reasoning: There are almost always season packs for older seasons, but they aren't always automatically found by Sonarr. This lets you avoid filling up your download client with single episodes for old seasons, by forcing a "season pack or bust" mentality.

Method: Adds an option in each Indexer's setting area for a maximum season age. If the air date of the episode is older than this set age, then only season packs are allowed.

#### Database Migration
Yes

markus101 commented on src/NzbDrone.Core/Datastore/Migration/170_add_MaximumSingleEpisodeAge_to_indexers.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(170)]
```
Please change to 172, 170 and 171 are already spoken for in work that is in progress.

markus101 reviewed (commented):
> There are almost always season packs for older seasons, but they aren't always automatically found by Sonarr. This lets you avoid filling up your download client with single episodes for old seasons, by forcing a "season pack or bust" mentality.

Why is it some season packs don't show up when searching?
Instead of something preventing non-season packs, can anything be done to solve this issue? 

Why should this be indexer specific setting? If it's set per indexer you can still run into the same underlying issue that you have right now, a single episode blocking a season pack from another indexer, especially if a higher allowed quality is grabbed in a single episode.

What about 2 episodes in a single file or anything that isn't a full season pack? If a double episode was released I think it'd be unexpected to have it grabbed over a season pack.

cjmanca commented:
> Why should this be indexer specific setting? If it's set per indexer you can still run into the same underlying issue that you have right now, a single episode blocking a season pack from another indexer, especially if a higher allowed quality is grabbed in a single episode.

My original PR had this as a global setting, but Taloth suggested changing it to per-indexer on discord (Talk > General channel a few hours ago)

> What about 2 episodes in a single file or anything that isn't a full season pack? If a double episode was released I think it'd be unexpected to have it grabbed over a season pack.

Originally I was going to restrict to full seasons only, but realized that some shows mostly just do numbered batches. Mainly anime though, so I suppose I could do full seasons for normal shows, and ">1 episodes" for anime.

> Why is it some season packs don't show up when searching?
> Instead of something preventing non-season packs, can anything be done to solve this issue?
They usually show up when doing an interactive search (but not always), but will either be rejected for various reasons ("Unknown series" is the main one, probably due to the naming scheme the release group used), or single episodes prioritized over the pack due to other settings (file size, quality, preferred words, etc). This is of course how Sonarr is designed to work, but personally I'd still rather get a few season packs (or nothing at all - in which case I'll manually find a season pack) over 1000 single episodes when adding a new series. If I don't like the season pack it picks, it's much easier to see which one and remove/replace manually. Much much harder to wade through hundreds or thousands of single episodes.

cjmanca commented:
Just to add though - prior to making this solution myself I looked for alternatives. The currently "official recommended" way on reddit that everyone is using seems to be adding regex filters to a tag which excludes all single episodes, and adding it to new series before searching, then removing it after getting the old seasons so it can pick up weeklies. This "tag dance" works when adding shows manually, but not for things like Ombi.

But since enough people do it, it felt like a feature that should just be a built in option.

cjmanca commented:
Ok, now updated based on feedback in Discord:

Now checks against the most recent episode in the season when comparing the air date, to prevent half a season being allowed and half not. A consequence of this is that it's for searches only now, not RSS.

No longer attempts to deal with anime or dailies, since they're inconsistent for batch/season packs. Only applies to Normal series.

Updated UI wording to better describe the setting and rejections related to it.

markus101 commented on src/NzbDrone.Core.Test/DecisionEngineTests/SingleEpisodeAgeDownloadDecisionFixture.cs:
```diff
@@ -0,0 +1,104 @@
+            multiSearch.Episodes = episodes.ToList();
+            multiSearch.SeasonNumber = 1;
+
+
```
Nit: Extra new line

markus101 commented on src/NzbDrone.Core.Test/DecisionEngineTests/SingleEpisodeAgeDownloadDecisionFixture.cs:
```diff
@@ -0,0 +1,104 @@
+                Release = new ReleaseInfo(),
+                ParsedEpisodeInfo = new ParsedEpisodeInfo { Quality = new QualityModel(Quality.SDTV, new Revision(version: 2)) },
+                Episodes = new List<Episode>()
+
```
Nit: Extra new line

markus101 commented on src/NzbDrone.Core.Test/DecisionEngineTests/SingleEpisodeAgeDownloadDecisionFixture.cs:
```diff
@@ -0,0 +1,104 @@
+                Episodes = new List<Episode>()
+
+            };
+
```
Nit: Extra new line

markus101 commented on src/NzbDrone.Core.Test/DecisionEngineTests/SingleEpisodeAgeDownloadDecisionFixture.cs:
```diff
@@ -0,0 +1,104 @@
+            };
+        }
+
+
```
Nit: Extra new line

markus101 commented on src/NzbDrone.Core.Test/DecisionEngineTests/SingleEpisodeAgeDownloadDecisionFixture.cs:
```diff
@@ -0,0 +1,104 @@
+            Subject.IsSatisfiedBy(parseResultSingle, multiSearch).Accepted.Should().Be(expectedResult);
+        }
+
+
```
Nit: Extra new line

markus101 commented on src/NzbDrone.Core/DecisionEngine/Specifications/SeasonPackOnlySpecification.cs:
```diff
@@ -0,0 +1,55 @@
+
+            if (searchCriteria.Episodes.Count == 1) return Decision.Accept();
+
+
```
Nit: Extra new line

markus101 commented on src/NzbDrone.Core/DecisionEngine/Specifications/SeasonPackOnlySpecification.cs:
```diff
@@ -0,0 +1,55 @@
+                }
+            }
+
+
```
Nit: Extra new line

markus101 commented on src/NzbDrone.Core.Test/DecisionEngineTests/SingleEpisodeAgeDownloadDecisionFixture.cs:
```diff
@@ -0,0 +1,104 @@
+
+
+
+
```
Nit: Many extra new lines

markus101 commented on frontend/src/Settings/Indexers/Indexers/EditIndexerModalContent.js:
```diff
@@ -153,6 +154,23 @@ function EditIndexerModalContent(props) {
+                <FormInputGroup
+                  type={inputTypes.NUMBER}
+                  name="maximumSingleEpisodeAge"
+                  helpText="During search, only season packs will be allowed when the season's most recent episode is older than this setting. Standard series only. Use 0 to ignore."
```
I think `Use 0 to disable.` reads a bit better.

markus101 replied:
Since this explicitly only triggers when a search for a season (or at least not a single episode only) is performed I think that should be more explicit.

cjmanca replied:
How about:
"During full season search, only season packs will be allowed when the season's most recent episode is older than this setting. Standard series only. Use 0 to disable."

markus101 replied:
Just a small tweak, but I think either:

`During a full season search...`
or
`During full season searches`

works a bit better.

cjmanca replied:
Done

markus101 commented on src/NzbDrone.Core/DecisionEngine/Specifications/SeasonPackOnlySpecification.cs:
```diff
@@ -0,0 +1,55 @@
+                return Decision.Accept();
+            }
+
+            if (searchCriteria.Episodes.Count == 1) return Decision.Accept();
```
Style wise I'd make this like `searchCriteria == null` and wrap the return in braces

markus101 commented on src/NzbDrone.Core/Indexers/IndexerBase.cs:
```diff
@@ -23,6 +23,7 @@ public abstract class IndexerBase<TSettings> : IIndexer
         public abstract string Name { get; }
         public abstract DownloadProtocol Protocol { get; }
         public int Priority { get; set; }
+        public int MaximumSingleEpisodeAge { get; set; }
```
Since this is more about season packs than single episodes I think this should be renamed to be more inline with that, I'm struggling a bit for a suggestion there though.

cjmanca replied:
Yeah, I had considered that too, but couldn't come up with a good wording to represent it. How about just adding "SeasonSearch" in front, like so:
"SeasonSearchMaximumSingleEpisodeAge"
?

markus101 replied:
Yeah, I like that.

markus101 reviewed (changes requested):
Sorry for the delay getting to this. Things are looking pretty good, but given the refactor to be more about rejecting single episodes during season searches I think the property should also be named to be inline with that.

markus101 commented on src/NzbDrone.Core/DecisionEngine/Specifications/SeasonPackOnlySpecification.cs:
```diff
@@ -0,0 +1,51 @@
+                if (subject.Series.SeriesType == SeriesTypes.Standard && !subject.ParsedEpisodeInfo.FullSeason && subject.Episodes.Count >= 1)
+                {
+                    // test against episodes of the same season in the current search, and make sure they have an air date
+                    var subset = searchCriteria.Episodes.Where(e => e.AirDateUtc.HasValue && e.SeasonNumber == subject.Episodes.First().SeasonNumber);
```
Add a `.ToList()` at the end so the `Count` and `Max` calls below don't recompute this twice.

markus101 commented on frontend/src/Settings/Indexers/Indexers/EditIndexerModalContent.js:
```diff
@@ -153,6 +154,23 @@ function EditIndexerModalContent(props) {
+                <FormInputGroup
+                  type={inputTypes.NUMBER}
+                  name="seasonSearchMaximumSingleEpisodeAge"
+                  helpText="During a full season search, only season packs will be allowed when the season's most recent episode is older than this setting. Standard series only. Use 0 to disable."
```
Reading this again I think the `,` after `During a full season search` is incorrect.

markus101 replied:
One other thing here would be to change `season's most recent episode` to `season's last episode` since this doesn't apply if the whole season hasn't aired.

markus101 commented on src/NzbDrone.Core/DecisionEngine/Specifications/SeasonPackOnlySpecification.cs:
```diff
@@ -0,0 +1,51 @@
+                    if (subset.Count() > 0 && subset.Max(e => e.AirDateUtc).Value.Before(DateTime.UtcNow - TimeSpan.FromDays(subject.Release.SeasonSearchMaximumSingleEpisodeAge)))
+                    {
+                        _logger.Debug("Release {0} aired more than {1} days ago, season pack required.", subject.Release.Title, subject.Release.SeasonSearchMaximumSingleEpisodeAge);
+                        return Decision.Reject("Aired more than {0} days ago, season pack required.", subject.Release.SeasonSearchMaximumSingleEpisodeAge);
```
One small tweak here, `Last episode in this season aired more than X days ago`

markus101 reviewed (changes requested):
Just a couple small things after I tested it out.

cjmanca commented:
Done

cjmanca commented:
@markus101  don't want to bug you, but it's been over a month, and I don't want it to be forgotten either

markus101 commented:
Needed to sort out some build issues and ship a main release before landing this, but that's been taken care of now. Thanks for working on this and your patience in getting it landed.
